src/lxc/conf.c

```diff
@@ -543,6 +543,11 @@ int lxc_rootfs_init(struct lxc_conf *conf, bool userns)
 			return syserror_set(-EINVAL, "Idmapped rootfs currently only supports the \"dir\" storage driver");
 	}
 
+	if (rootfs->bdev_type && strequal(rootfs->bdev_type, "zfs")) {
+		TRACE("Not pinning because container uses ZFS");
+		goto out;
+	}
```

Comment:
Hm, actually this should be slighlty different. You should move this into the `if (rootfs->path)` condition below where overlayfs is already checked:
```c
	if (rootfs->path) {
		if (rootfs->bdev_type) {
	                if (strequal(rootfs->bdev_type, "overlay") || strequal(rootfs->bdev_type, "overlayfs")))
			        return log_trace_errno(0, EINVAL, "Not pinning on stacking filesystem");
                        if (strequal(rootfs->bdev_type, "zfs"))
                                return log_trace_errno(0, EINVAL, "Not pinning on ZFS filesystem");
	        }
```

Reply:
Agreed, that reads much better. Pushed the change now.